ecs_deploy/ecs.py

```diff
@@ -473,6 +595,267 @@ def apply_container_secrets(self, container, new_secrets, exclusive=False):
+
+    def set_port_mappings(self, port_mappings_list, exclusive=False):
+        port_mappings = {}
+
+        for port_mapping in port_mappings_list:
+            port_mappings.setdefault(port_mapping[0], [])
+            mapping = {}
+            mapping["containerPort"] = int(port_mapping[1])
+            mapping["hostPort"] = int(port_mapping[2])
+            mapping["protocol"] = "tcp"
+            port_mappings[port_mapping[0]].append(mapping)
+
+        self.validate_container_options(**port_mappings)
+        for container in self.containers:
+            if container[u'name'] in port_mappings:
+                self.apply_container_port_mappings(
+                    container=container,
+                    new_port_mappings=port_mappings[container[u'name']],
+                    exclusive=exclusive,
+                )
+            elif exclusive is True:
+                self.apply_container_port_mappings(
+                    container=container,
+                    new_port_mappings={},
+                    exclusive=exclusive,
+                )
+
+    def apply_container_port_mappings(self, container, new_port_mappings, exclusive=False):
+        port_mappings = container.get('portMappings', [])
+        old_port_mappings = [ {name: value for name, value in port_mapping.items()} for port_mapping in port_mappings]
+
+        if exclusive is True:
+            merged = new_port_mappings if new_port_mappings else []
+        else:
+            merged = copy.deepcopy(old_port_mappings)
+            old_container_ports = [port_mapping["containerPort"] for port_mapping in merged]
+            new_container_ports = [port_mapping["containerPort"] for port_mapping in new_port_mappings]
+            new = set(new_container_ports) - set(old_container_ports)
+            new_merged = list()
+            # update
+            for port_mapping in merged:
+                for new_port_mapping in new_port_mappings:
+                    if port_mapping["containerPort"] == new_port_mapping["containerPort"]:
+                        port_mapping.update(new_port_mapping)
+            # new
+            for port in new:
+                for new_port_mapping in new_port_mappings:
+                    if port == new_port_mapping["containerPort"]:
+                        new_merged.append(new_port_mapping)
+
+            merged.extend(new_merged)
+        if old_port_mappings == merged:
+            return
+
+        diff = EcsTaskDefinitionDiff(
+            container=container[u'name'],
+            field=u'portMappings',
+            value=merged,
+            old_value=old_port_mappings
+        )
+        self._diff.append(diff)
+
+        container[u'portMappings'] = [
+            {name: value for name,value in e.items()} for e in merged
+        ]
+
+    def set_mount_points(self, mount_points_list, exclusive=False):
+        mount_points = {}
+        for mount_point in mount_points_list:
+            mount_points.setdefault(mount_point[0], [])
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

You are using the `dict.setdefault()` method. We recommend using the `defaultdict` object to handle missing items in internal state because the default value is automatically generated when you access or modify a missing key. `defaultdict` can also be easier to use and understand.

[Learn more](https://docs.python.org/3/library/collections.html#collections.defaultdict)

Reply:
Should be fine now.